app/src/main/java/com/android/sample/ui/authentication/SignIn.kt

```diff
@@ -58,31 +59,19 @@ fun SignInScreen(
 ) {
   val context = LocalContext.current
   val scope = rememberCoroutineScope()
+  var showLoadingDialog by remember { mutableStateOf(false) }
+
   val user by remember { mutableStateOf(Firebase.auth.currentUser) }
   val userAccount by userAccountViewModel.userAccount.collectAsState(initial = null)
 
   val launcher =
       rememberFirebaseAuthLauncher(
           onAuthComplete = { result ->
+            showLoadingDialog = true // Show loading screen
             Log.d("SignInScreen", "User signed in: ${result.user?.displayName}")
-            val userId = result.user?.uid
-            if (userId != null) {
-              userAccountViewModel.getUserAccount(userId)
-
-              scope.launch {
-                delay(450) // delay introduced to wait for data to be fetched
-                // TODO: Find a way to modify this with loading screen
-
-                // Observe changes in userAccount to know if profile exists
-                userAccountViewModel.userAccount.collect { account ->
-                  if (account != null) {
-                    // If account exists, navigate to main screen
-                    navigationActions.navigateTo(TopLevelDestinations.MAIN)
-                  } else {
-                    // If no account exists, navigate to AddAccount screen
-                    navigationActions.navigateTo(Screen.ADD_ACCOUNT)
-                  }
-                }
+            result.user?.uid?.let { userId ->
+              checkUserAccountAndNavigate(userId, userAccountViewModel, navigationActions, scope) {
```

Comment:
Really good practice to write a new function for readability and avoid duplicating code